core/server/api_container/server/service_network/default_service_network.go

```diff
@@ -1791,3 +1878,102 @@ func renderTemplateToFile(templateAsAString string, templateData interface{}, de
 	}
 	return nil
 }
+
+func waitUntilAllTCPAndUDPPortsAreOpen(
+	ipAddr net.IP,
+	ports map[string]*port_spec.PortSpec,
+) error {
+
+	var portCheckErrorGroup errgroup.Group
+
+	for _, portSpec := range ports {
+		wrappedWaitFunc := func() error {
+			return waitUntilPortIsOpenWithTimeout(ipAddr, portSpec)
+		}
+		portCheckErrorGroup.Go(wrappedWaitFunc)
+	}
+	//This error group pattern allow us to reject early if at least one of the ports check fails
+	//in this opportunity we want to reject early because the main user pain that we want to handle
+	//is a wrong service configuration
+	if err := portCheckErrorGroup.Wait(); err != nil {
+		return stacktrace.Propagate(err, "An error occurred while waiting for all TCP and UDP ports to be open")
+	}
+
+	return nil
+}
+
+func waitUntilPortIsOpenWithTimeout(
+	ipAddr net.IP,
+	portSpec *port_spec.PortSpec,
+) error {
+	// reject early if it's disable
+	if !portSpec.GetWait().GetEnable() {
+		return nil
+	}
+	timeout := portSpec.GetWait().GetTimeout()
+
+	var (
+		ticker                  = time.NewTicker(timeout)
+		startTime               = time.Now()
+		finishTime              = startTime.Add(timeout)
+		shouldContinueInTheLoop = true
+		retries                 = 0
+		err                     error
+	)
+
+	time.Sleep(portSpec.GetWait().GetInitialDelay())
+
+	for shouldContinueInTheLoop {
+		select {
+		case <-ticker.C:
+			return stacktrace.NewError("Scanning ports has reached the '%v' time out", timeout.String())
+		default:
+			now := time.Now()
+			scanPortTimeout := finishTime.Sub(now)
+			if err = scanPort(ipAddr, portSpec, scanPortTimeout); err == nil {
+				shouldContinueInTheLoop = false
+				break
+			}
+			time.Sleep(time.Duration(waitForPortsOpenRetriesDelayMilliseconds) * time.Millisecond)
```

Comment:
nit: you don't need the `time.Duration` here, something like `time.Sleep(waitForPortsOpenRetriesDelayMilliseconds * time.Millisecond)` should work, I think

Reply:
I've removed this part, I following your suggested pattern instead